### DISCUSSION TITLE
Trivy Image Analysis Log Output differs from CDX Format

### DISCUSSION THREAD

**Discussion by mymichu** [aquasecurity/trivy#4819] [Q&A] [answered]
### Question

We (@rplessl and me) discovered that Trivy analysis has a different output vulnerability rating on the stdout than if we executed the analysis tool with the output format cyclonedx. For example, the vulnerability "CVE-2022-39327" is rated in the stdout as High and the cyclonedx output as "Critical" and "High". It is unclear whether this is a bug or if Trivy goes with the lowest vulnerability rating if two ratings are available. Please clarify to us what the Trivy approach is. 

Prerequisite:

`${TRIVY_DB_REPOSITORY:=--db-repository ghcr.io/aquasecurity/trivy-db:2 --java-db-repository: ghcr.io/aquasecurity/trivy-java-db:1 }"`

If we execute Trivy with the following command:

`trivy image ${TRIVY_DB_REPOSITORY} --severity HIGH,CRITICAL --no-progress ${IMAGE}`

We get the following stdout:

`
Total: 1 (HIGH: 1, CRITICAL: 0)
┌──────────────────────┬────────────────┬──────────┬───────────────────┬───────────────┬──────────────────────────────────────────────────────────────┐
│       Library        │ Vulnerability  │ Severity │ Installed Version │ Fixed Version │                            Title                             │
├──────────────────────┼────────────────┼──────────┼───────────────────┼───────────────┼──────────────────────────────────────────────────────────────┤
│ azure-cli (METADATA) │ CVE-2022-39327 │ HIGH     │ 2.36.0            │ 2.40.0        │ Improper Control of Generation of Code ('Code Injection') in │
│                      │                │          │                   │               │ Azure CLI                                                    │
│                      │                │          │                   │               │ https://avd.aquasec.com/nvd/cve-2022-39327                   │
└──────────────────────┴────────────────┴──────────┴───────────────────┴───────────────┴──────────────────────────────────────────────────────────────┘`

If we execute Trivy with the following command:

`trivy image ${TRIVY_DB_REPOSITORY} --severity HIGH,CRITICAL --quiet --format cyclonedx ${IMAGE} --scanners vuln`

We get the following cyclonedx output "snippet":
```
{
      "id": "CVE-2022-39327",
      "source": {
        "name": "ghsa",
        "url": "https://github.com/advisories?query=type%3Areviewed+ecosystem%3Apip"
      },
      "ratings": [
        {
          "source": {
            "name": "ghsa"
          },
          "score": 8.1,
          "severity": "high",
          "method": "CVSSv31",
          "vector": "CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:U/C:H/I:H/A:H"
        },
        {
          "source": {
            "name": "nvd"
          },
          "score": 9.8,
          "severity": "critical",
          "method": "CVSSv31",
          "vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H"
        }
      ],
      "cwes": [
        78
      ],
      "description": "Azure CLI is the command-line interface for Microsoft Azure. In versions previous to 2.40.0, Azure CLI contains a vulnerability for potential code injection. Critical scenarios are where a hosting machine runs an Azure CLI command where parameter values have been provided by an external source. The vulnerability is only applicable when the Azure CLI command is run on a Windows machine and with any version of PowerShell and when the parameter value contains the `\u0026` or `|` symbols. If any of these prerequisites are not met, this vulnerability is not applicable. Users should upgrade to version 2.40.0 or greater to receive a a mitigation for the vulnerability.",
      "recommendation": "Upgrade azure-cli to version 2.40.0",
      "advisories": [
        {
          "url": "https://github.com/Azure/azure-cli/pull/23514"
        },
        {
          "url": "https://github.com/Azure/azure-cli/pull/24015"
        },
        {
          "url": "https://github.com/Azure/azure-cli/security/advisories/GHSA-47xc-9rr2-q7p4"
        },
        {
          "url": "https://github.com/advisories/GHSA-47xc-9rr2-q7p4"
        },
        {
          "url": "https://nvd.nist.gov/vuln/detail/CVE-2022-39327"
        }
      ],
      "published": "2022-10-25T17:15:00+00:00",
      "updated": "2023-06-27T17:20:00+00:00",
      "affects": [
        {
          "ref": "pkg:pypi/azure-cli@2.36.0?file_path=usr%2Flocal%2Flib%2Fpython3.10%2Fsite-packages%2Fazure_cli-2.36.0.dist-info%2FMETADATA",
          "versions": [
            {
              "version": "2.36.0",
              "status": "affected"
            }
          ]
        }
      ]
    },
```

### Target

Container Image

### Scanner

Vulnerability

### Output Format

None

### Mode

Standalone

### Operating System

Linux alpine:3.18.2

### Version

```bash
Version: 0.43.1
```

**Comment by knqyf263** [comment, marked as answer]
The stdout shows the most appropriate severity only. But CycloneDX shows severities from all the data sources. You can see the same in the JSON format.
https://aquasecurity.github.io/trivy/v0.43/docs/scanner/vulnerability/os/#data-source-selection

**Reply by mymichu** [reply]
I appreciate your help. This was very helpful.
